File: OmegaUp/1357. Carretera.cpp
#include <iostream>
#include <vector>
using namespace std;

struct Car {
  int x;
  int v;
};

double distance ( vector<Car> car, double time ) {
  double distance_max = -1e5;
  double distance_min = 1e5;
  for ( int i = 0; i < car.size(); i++ ) {
    double distance_car = (double) car[i].x + car[i].v * time;
    distance_max = max ( distance_car, distance_max );
    distance_min = min ( distance_car, distance_min );
  }
  return distance_max - distance_min;
}

double func ( vector<Car> car ) {
  const double EPSILON = 1e-7;
  double max = 1e5;
  double min = 0;
  while ( max - min >= EPSILON ) {
    double time1 = min + ( max - min ) / 3;
    double time2 = max - ( max - min ) / 3;

    double distance_1 = distance ( car, time1 );
    double distance_2 = distance ( car, time2 );
    
    if ( distance_1 < distance_2 )
      max = time2;
    else 
      min = time1;
  }
  return distance ( car, min );
}

int main () {
  int n;
  cin >> n;
  vector<Car> cars(n); 
  while ( n-- ) {
    int x, v;
    cin >> x;
    cin >> v;
    Car car;
    car = { x, v };
    cars[n] = car;
  }
  printf ( "%.7f", func ( cars ) );
}

File: OmegaUp/1357. Carretera.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;

public class Main {
    static double dis(int[][] cars, double time) {
        double max = -1e5;
        double min = 1e5;
        for (int i = 0; i < cars[0].length; i++) {
            double pos = (double)cars[0][i] + cars[1][i] * time;
            max = Math.max(max, pos);
            min = Math.min(min, pos);
        }
        return max - min;
    }

    static double func(int[][] cars) {
        double low = 0;
        double high = 1e5;
        final double EPSILON = 1e-6;
        while (high - low >= EPSILON) {
            double mid1 = low + (high - low) / 3;
            double mid2 = high - (high - low) / 3;

            double distance1 = dis(cars, mid1);
            double distance2 = dis(cars, mid2);

            if (distance1 < distance2)
                high = mid2;
            else
                low = mid1;
        }
        return dis(cars, low);
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] firstLine = br.readLine().split(" ");
        int n = Integer.parseInt(firstLine[0]);
        int[][] cars = new int[2][n];

        for (int i = 0; i < n; i++) {
            String[] carData = br.readLine().split(" ");
            cars[0][i] = Integer.parseInt(carData[0]);
            cars[1][i] = Integer.parseInt(carData[1]); 
        }

        System.out.printf("%.6f", func(cars));
    }
}
